controllers/drplacementcontrol.go

```diff
@@ -0,0 +1,1372 @@
+
+	const restorePVs = true
+
+	result, err := d.executeRelocation(newHomeCluster, "", restorePVs)
+	if err != nil {
+		rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeWarning,
+			rmnutil.EventReasonSwitchFailed, err.Error())
+
+		return !done, err
+	}
+
+	d.advanceToNextDRState()
+	d.log.Info("Exiting runFailover", "state", d.getLastDRState())
+	setMetricsTimerFromDRState(rmn.FailingOver, timerStop)
+
+	return result, nil
+}
+
+func (d *DRPCInstance) getCurrentHomeClusterName() string {
+	curHomeCluster := ""
+	if len(d.userPlacementRule.Status.Decisions) > 0 {
+		curHomeCluster = d.userPlacementRule.Status.Decisions[0].ClusterName
+	}
+
+	if curHomeCluster == "" {
+		curHomeCluster = d.instance.Status.PreferredDecision.ClusterName
+	}
+
+	return curHomeCluster
+}
+
+func (d *DRPCInstance) runRelocate() (bool, error) {
+	d.log.Info("Entering runRelocate", "state", d.getLastDRState())
+
+	const done = true
+
+	if d.isFailoverInProgress() {
+		return done, fmt.Errorf("invalid state %s for the selected action %v",
+			d.getLastDRState(), d.instance.Spec.Action)
+	}
+
+	preferredCluster := d.instance.Spec.PreferredCluster
+	preferredClusterNamespace := preferredCluster
+	// We are done if empty. Relocation requires preferredCluster to be configured
+	if preferredCluster == "" {
+		return !done, fmt.Errorf("preferred cluster not valid")
+	}
+
+	if preferredCluster == d.instance.Spec.FailoverCluster {
+		return done, fmt.Errorf("failoverCluster and preferredCluster can't be the same %s/%s",
+			d.instance.Spec.FailoverCluster, preferredCluster)
+	}
+
+	// We are done if already failed back
+	if d.hasAlreadySwitchedOver(preferredCluster) {
+		err := d.ensureCleanup(preferredCluster)
+		if err != nil {
+			return !done, err
+		}
+
+		return done, nil
+	}
+
+	if len(d.vrgs) == 0 {
+		return !done, fmt.Errorf("no VRGs exists. Can't relocate")
+	}
+
+	// For relocation, the current home cluster is not the preferredCluster
+	if !d.isCurrentHomeClusterReadyForRelocation(preferredCluster) {
+		d.log.Info("Waiting for current home cluster to be ready for relocation")
+
+		return !done, nil
+	}
+
+	return d.relocate(preferredCluster, preferredClusterNamespace, rmn.Relocating)
+}
+
+func (d *DRPCInstance) relocate(preferredCluster, preferredClusterNamespace string, drState rmn.DRState) (bool, error) {
+	const done = true
+	// Make sure we record the state that we are failing over
+	d.setDRState(drState)
+	setMetricsTimerFromDRState(drState, timerStart)
+
+	// Setting up relocation insures that all VRGs in all managed cluster are secondaries
+	err := d.setupRelocation(preferredCluster)
+	if err != nil {
+		return !done, err
+	}
+
+	// After we insured that all clusters are secondaries, we then insure
+	// that the current home cluster is ready for relocation.
+	// For relocation, the current home cluster is not the preferredCluster
+	if !d.isCurrentHomeClusterReadyForRelocation(preferredCluster) {
```

Comment:
This check is made in the above function runRelocate (from where this function i.e. "d.relocate" is invoked"). Should this check be made again here?

Reply:
Yes, I know.  This double call is confusing and I think it can be avoided if I put a little bit more thought into it. But for now, it is necessary.  The first call waits for the current cluster to be in a ready state, the second call happens after the current cluster is guaranteed to be secondary.